<robot name="my_SCARA">
    <!-- defin links -->
    <link name="base_link"/>
    <link name="stand_link">
        <visual>
            <geometry>
                <box size="0.7 0.3 0.3"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <material name="white">
                <color rgba="1.0 1.0 1.0 1.0"/>
            </material>
        </visual>
    </link>
    <link name="arm1_link">
        <visual>
            <geometry>
                <box size="0.7 0.1 0.1" />
            </geometry>
            <origin xyz="0.3 0 0" rpy="0 0 0" />
            <material name="blue">
                <color rgba="0.0 0.0 1.0 1.0" />
            </material>
        </visual>
    </link>
    <link name="arm2_link">
        <visual>
            <geometry>
                <box size="0.35 0.1 0.1" />
            </geometry>
            <origin xyz="0.125 0 0" rpy="0 0 0" />
            <material name="green">
                <color rgba="0.0 1.0 0.0 1.0" />
            </material>
        </visual>
    </link>
    <link name="linear_link">
        <visual>
            <geometry>
                <box size="0.1 0.1 0.4" />
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <material name="orange">
                <color rgba="1.0 0.5 0.0 1.0" />
            </material>
        </visual>
    </link>
     <link name="wrist_link">
        <visual>
            <geometry>
                <box size="0.3 0.1 0.04" />
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <material name = "white">
                <color rgba="1.0 1.0 1.0 1.0" />
            </material>
        </visual>
    </link>
    <link name="righthand_link">
        <visual>
            <geometry>
                <cylinder radius="0.05" length="0.01"/>
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <material name="red">
                <color rgba="1.0 0.0 0.0 1.0" />
            </material>
        </visual>
    </link>
    <link name="lefthand_link">
        <visual>
            <geometry>
                <cylinder radius="0.05" length="0.01" />
            </geometry>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <material name="red">
                <color rgba="1.0 0.0 0.0 1.0" />
            </material>
        </visual>
    </link>

    <!-- define joints -->
    <joint name="base_stand" type="fixed">
        <parent link="base_link" />
        <child link="stand_link" />
        <origin xyz="0 0 0.15" rpy="0 0 0" />
    </joint>
    <joint name="stand_arm1" type="revolute">
        <parent link="stand_link"/>
        <child link="arm1_link"/>
        <origin xyz="0 0 0.2" rpy="0 0 0"/>
        <limit lower="-3.1416" upper="3.1416" effort="10" velocity="10"/>
        <joint_properties damping="0.0" friction="0.0" />
        <axis xyz="0 0 1"/>
    </joint>
    <joint name="arm1_arm2" type="revolute">
        <parent link="arm1_link"/>
        <child link="arm2_link"/>
        <origin xyz="0.6 0 0.1" rpy="0 0 0" />
        <limit lower="-3.1416" upper="3.1416" effort="10" velocity="10" />
        <joint_properties damping="0.0" friction="0.0" />
        <axis xyz="0 0 1" />
    </joint>
    <joint name="arm2_linear" type="prismatic">
        <parent link="arm2_link"/>
        <child link="linear_link"/>
        <origin xyz="0.35 0 0" rpy="0 0 0"/>
        <limit lower="-0.15" upper="0.15" effort="10" velocity="10"/>
        <axis xyz="0 0 1"/>
    </joint>
    <joint name="linear_wrist" type="revolute">
        <parent link="linear_link"/>
        <child link="wrist_link"/>
        <origin xyz="0 0 -0.22" rpy="0 0 0"/>
        <limit lower="-1.57075" upper="1.57075" effort="10" velocity="10"/>
        <axis xyz="0 0 1"/>
    </joint>
    <joint name="wrist_right" type="fixed">
        <parent link="wrist_link"/>
        <child link="righthand_link"/>
        <origin xyz="0.1 0 -0.025" rpy="0 0 0"/>
    </joint>
    <joint name="wrist_left" type="fixed">
        <parent link="wrist_link" />
        <child link="lefthand_link" />
        <origin xyz="-0.1 -0 -0.025" rpy="0 0 0" />
    </joint>
 
    <transmission name="arm1_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="stand_arm1">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="arm1_motor">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanismReduction>1</mechanismReduction>
        </actuator>
    </transmission>
    <transmission name="arm2_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="arm1_arm2">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="arm2_motor">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanismReduction>1</mechanismReduction>
        </actuator>
    </transmission>
    <transmission name="linear_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="arm2_linear">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="linear_motor">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanismReduction>1</mechanismReduction>
        </actuator>
    </transmission>
    <transmission name="wrist_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="linear_wrist">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="wrist_motor">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanismReduction>1</mechanismReduction>
        </actuator>
    </transmission>

    <!-- make gazebo plugin -->
    <gazebo reference="stand_link">
        <material>Gazebo/White</material>
    </gazebo>
    <gazebo reference="arm1_link">
        <material>Gazebo/Blue</material>
    </gazebo>
    <gazebo reference="arm2_link">
        <material>Gazebo/Green</material>
    </gazebo>
    <gazebo reference="linear_link">
        <material>Gazebo/Orange</material>
    </gazebo>
    <gazebo reference="wrist_link">
        <material>Gazebo/White</material>
    </gazebo>
    <gazebo reference="righthand_link">
        <material>Gazebo/Red</material>
    </gazebo>
    <gazebo reference="lefthand_link">
        <material>Gazebo/Red</material>
    </gazebo>
    <gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
            <robotNamespace>/my_SCARA</robotNamespace>
            <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
            <legacyModeNS>true</legacyModeNS>
        </plugin>
    </gazebo>
</robot>
